### PR TITLE
Desktop: Fixes #5178: NoteListItem's height can be properly styled using CSS for v2.8+

### LINKED ISSUE #5178: Incorrect number of note items and unused dead area in Note List panel
To increase the displayed information of Joplin, I reduced the height of note items in the Note List using the following userchrome.css. 

    .sidebar div.list-item-container {
        height: 20px !important;
    }

    .note-list .list-item-container,
    .note-list .list-item-container div {
        height: 25px !important;
    }

The result is as follows. <strong>The bottom area of the Note List is unused</strong>, and <strong>the number of the notes in the list is unchanged before and after the height is changed</strong>. For comparison, the Sidebar is properly rendered. 

joplin_bug_20210711_2a

For reference, the before image is here. 

joplin_bug_20210711_1

## Environment

Joplin version: 2.1.8
Platform: Windows
OS specifics: Windows 10 Pro (21H1)

## Steps to reproduce

1. Run with the above userchrome.css.
2. Click 'All notes' in the Sidebar. If you have enough number of notes, you always observe this bug in the Note List panel.

## Describe what you expected to happen

When the height of items in the Note List is reduced, the whole area should be used and the number of items should be increased as follows.

joplin_bug_20210711_expected

### PR DESCRIPTION
**Pull request by ken1kob**
The bug reported in issue #5178 is fixed.
This PR is the replacement of PR #5271 and is for the refactored NoteList function component in v2.8+.

The following spec is the same with the before, but its implementation is more effecient, because it's based on the efficiency of the React function component.

### Specification

When the height of NoteListItem is styled using `userchrome.css` as below, the number of displayed note items is correctly calcurated, and there is no unused dead area in the NoteList panel.

    .note-list .list-item-container,
    .note-list .list-item-container div {
        height: 25px !important;
    }

By applying this fix, you see what people expects as below shown in issue #5178.

joplin_bug_20210711_expected

### CODE REVIEW

File: packages/app-desktop/gui/NoteList/NoteList.tsx [hunk 1]
```diff
@@ -453,6 +453,14 @@ const NoteListComponent = (props: Props) => {
 		};
 	}, []);
 
+	useEffect(() => {
+		const noteItem = Object.values<any>(itemAnchorRefs_.current)[0]?.current;
+		const actualItemHeight = noteItem?.getHeight() ?? 0;
+		if (actualItemHeight >= 8) { // To avoid generating too many narrow items
+			setItemHeight(actualItemHeight);
+		}
+	});
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteList/NoteList.tsx]: Please add a comment to explain why this is needed

Comment (ken1kob) [reply packages/app-desktop/gui/NoteList/NoteList.tsx]: Done.

**Comment by Daeraxa** [comment]
Somebody on discord found a bug in the dev version and I think I've traced it back to this commit.
Replication steps:
- Enable custom sorting on a notebook
- Drag a note to reorder
-> Notelist items shrink

image

Size is restored on restart

**Comment by laurent22** [comment]
I can replicate this too

**Comment by ken1kob** [comment]
> * Drag a note to reorder
>       -> Notelist items shrink

If you create an issue, I can start working on it.

**Comment by Daeraxa** [comment]
Sure, just wanted to make you aware in case I had overlooked something

**Comment by antoniopaolini** [comment]
Not works anymore. Any way to have more compact notebook list?

**Pull request merged**